## Question 1

Define

$$
x = [-9, 5, 1, -4, 12, -7]^T
$$

and calculate the following norms:

(a) $\|x\|_1$ by hand and using Python.

(b) $\|x\|_2$ by hand and using Python.

(c) $\|x\|_\infty$ by hand and using Python.

You may need to import:

```python
from numpy import inf, array
from scipy.linalg import norm
```


In [2]:
! pip install scipy

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.3/62.3 kB 2.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 20.4/20.4 MB 4.7 MB/s eta 0:00:0000:0100:01

[notice] A new release of pip is available: 24.0 -> 26.1.2
[notice] To update, run: pip install --upgrade pip


In [3]:

from numpy import inf, array
from scipy.linalg import norm

In [7]:
x = array([-9 , 5, 1, -4, 12, -7])

x_1 = norm(x, 1)
x_2 = norm(x, 2)
x_inf = norm(x, inf)

print(x_1)
print(x_2)
print(x_inf)

38.0
17.776388834631177
12.0


## Question 2

It can be shown that for $x \in \mathbb{R}^n$,

$$
\|x\|_2 \leq \|x\|_1 \leq \sqrt{n}\|x\|_2,
$$

$$
\|x\|_\infty \leq \|x\|_2 \leq \sqrt{n}\|x\|_\infty,
$$

$$
\|x\|_\infty \leq \|x\|_1 \leq n\|x\|_\infty.
$$

Find nonzero vector(s) to show that these inequalities are tight.


In [8]:
x = array([1, 0, 0])
x_1 = norm(x, 1)
x_2 = norm(x, 2)
x_inf = norm(x, inf)

print(x_1)
print(x_2)
print(x_inf)

1.0
1.0
1.0


## Question 3

Consider the matrices $A \in \mathbb{R}^{5 \times 4}$ and $B \in \mathbb{R}^{4 \times 4}$ given by

$$
A =
\begin{bmatrix}
3 & -2 & 0 & 5 \\
1 & 8 & 7 & 0 \\
-9 & 5 & 1 & 2 \\
3 & -2 & 0 & 5 \\
3 & -4 & 0 & 5
\end{bmatrix},
\qquad
B =
\begin{bmatrix}
3 & -2 & 0 & 5 \\
1 & 8 & 7 & 0 \\
-9 & 5 & 1 & 2 \\
3 & -2 & 0 & 5
\end{bmatrix}.
$$

(a) For the matrix $A$, calculate:

(i) $\|A\|_1$ by hand and using Python.

(ii) $\|A\|_\infty$ by hand and using Python.

(iii) $\|A\|_2$ using Python.

(b) For the matrix $B$:

(i) How can it be efficiently calculated from the matrix $A$?

(ii) Calculate $\|B\|_1$, $\|B\|_\infty$, and $\|B\|_2$.

(iii) Calculate the largest magnitude eigenvalue $|\lambda_{\max}|$ of $B$. Does $\|B\|_2$ equal $|\lambda_{\max}|$?

You may need to import:

```python
from numpy import inf, array
from scipy.linalg import norm, eigvals
```


In [ ]:
from numpy import inf, array
from scipy.linalg import norm, eigvals
# a)
A = array([
    [3, -2, 0, 5],
    [1, 8, 7, 0],
    [-9, 5, 1, 2],
    [3, -2, 0, 5],
    [3, -4, 0, 5]
])
# i) 
A_1 = norm(A, 1)

# ii)
A_inf = norm(A, inf)

# iii)
A_2 = norm(A, 2)

print(A_1)
print(A_2)
print(A_inf)

# b)
# i) 
B = A[0:-1]

# ii) 
B_1 = norm(B, 1)
B_inf = norm(B, inf)
B_2 = norm(B, 2)

# iii)
eigs = eigvals(B)
eig_max = max(abs(eigs))
print(B_1)
print(B_2)
print(B_inf)
print(eig_max)
print(eigs)

21.0
14.138407784747079
17.0
17.0
12.965395613366699
17.0
(12.495863904919737+0j)
[ 1.24958639e+01+0.j -1.53671568e+00+0.j  3.84371012e-15+0.j
  6.04085178e+00+0.j]


## Question 4

For the following matrices, calculate the condition numbers with respect to the $1$-, $2$-, and $\infty$-norms.

(a) $A = \alpha I$, where $\alpha \in \mathbb{R}$, $\alpha \neq 0$, and $I$ is the $n \times n$ identity matrix.

(b) The $n \times n$ Hilbert matrix $H = [h_{ij}]$ with entries

$$
h_{ij} = \frac{1}{i + j - 1},
\qquad
 i,j = 1, \ldots, n,
$$

in the cases $n = 5, 10, 15$.

You may need to import:

```python
from numpy import inf, identity, linalg
from scipy.linalg import hilbert
```


In [ ]:
from numpy import inf, identity, linalg
from scipy.linalg import hilbert

# a) 


## Question 5

Consider solving a linear system $Ax = b$ where the matrix $A$ is known exactly.

(a) Estimate the number of significant figures in the computed solution when the right-hand-side vector $b$ is known to 8 significant figures and:

(i) $\|A\| = 1.2 \times 10^1$, $\|A^{-1}\| = 3.9 \times 10^3$.

(ii) $A$ is symmetric and has ordered eigenvalues $0.00031, 0.0045, \ldots, 101.4, 198.2$.

(iii) The reciprocal condition number of $A$ is $9.9010 \times 10^{-12}$.

(b) In each case, what must the relative error in $b$ be to get a computed solution with 6 significant figures?

You may need to import:

```python
from numpy import finfo, log10, floor
```


## Question 6

Write a Python function `symchk(A, tol)` in a `symchkSol.py` file to determine if a matrix

$$
A = [a_{ij}] \in \mathbb{R}^{m \times n}
$$

is symmetric, that is, if $A = A^T$.

Start the `symchkSol.py` file with:

```python
from numpy import maximum, any, finfo, float32, float64, ndarray

def symchk(A, tol=None):
```

Notes:

- The function returns `True` when the input matrix is symmetric to within a relative tolerance `tol`, in the sense that

$$
|a_{ij} - a_{ji}| \leq \mathrm{tol}\max\{|a_{ij}|, |a_{ji}|\}
$$

for all $i$ and $j$, and returns `False` otherwise.

- If the matrix is not square, then it cannot be symmetric.

- If the input argument `tol` is not specified, use the default value $10n^2\epsilon$, where $\epsilon$ is the machine epsilon.

- If the input argument `tol` is negative, raise a `ValueError`.

- Create a Python file to test your function `symchk` on a variety of inputs.

You may need to import:

```python
from symchkSol import symchk
from numpy import finfo, array, inf
from numpy.random import rand, randn
from scipy.linalg import norm
```
